### PR TITLE
ear-1387-update-publisher-v3-for-question-flow

### PR DESCRIPTION
**Pull request by jchapman68**
Added output for questionnaire_flow.
Some extra work was done as the validator was throwing errors around empty attributes.

### CODE REVIEW

File: src/eq_schema/schema/Answer/index.js [hunk 1]
```diff
@@ -19,9 +19,15 @@ class Answer {
...
-    this.description = answer.description;
 
+    if (answer.label) {
+      this.label = answer.label;
+    }
+
+    if (answer.description) {
+      this.description = answer.description;
+    }
+
```
Comment (MattElcock) [inline src/eq_schema/schema/Answer/index.js:30]: How come this is here? Doesn't look like it's necessary for this ticket, or is it?

File: src/eq_schema/schema/Group/index.js [hunk 2]
```diff
@@ -8,7 +8,9 @@ const translateAuthorSkipconditions = require("../../builders/skipConditions");
 class Group {
   constructor(title, folder, ctx) {
     this.id = `group${folder.id}`;
-    this.title = ctx.questionnaireJson.navigation ? title : "";
+    if (title) {
+      this.title = title;
+    }
```
Comment (MattElcock) [inline src/eq_schema/schema/Group/index.js:13]: How come this is here? Doesn't look like it's necessary for this ticket, or is it?

File: src/eq_schema/schema/Questionnaire/index.js [hunk 3]
```diff
@@ -62,22 +53,14 @@ class Questionnaire {
     };
   }
 
-  buildHub(hub) {
-    if (hub) {
-      return new Hub(hub);
-    }
-    return { enabled: false };
+  buildQuestionnaire_Flow(questionnaireJson) {
```
Comment (MattElcock) [inline src/eq_schema/schema/Questionnaire/index.js:56]:
```suggestion
  buildQuestionnaireFlow(questionnaireJson) {
```

**Review by MattElcock — CHANGES_REQUESTED** [review]
Hi Josh,

You've used a hybrid of snake and camel case in your function names (`buildQuestionnaire_Flow`). Could you run back through your code and rewrite them in camel case (`buildQuestionnaireFlow`) instead, please? 

Your `Questionnaire_Flow` filename, and the functions within that class, should be camel case, as well.

Then, it'll be consistent with the rest of the app and will be in line with our code standards.

You also seem to have included work that isn't covered by the ticket. I've hightlighted some of them. Could you expand on why they've been changed, please? If they're necessary, it might also be worth including the reasoning in the PR description to remind reviewers to test those changes as well.

File: src/eq_schema/schema/Group/index.js (hunk 2 above)
Comment (martyncolmer) [reply src/eq_schema/schema/Group/index.js:13]: This was added as the validator didn't like empty titles.

**Review by martyncolmer — DISMISSED** [review]

File: src/eq_schema/schema/QuestionnaireFlow/index.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  build_required_sections(questionnaireJson){
+    const sections = filter(questionnaireJson.sections, {"required": true});
+    const required_sections = sections.map((section) => section.id);
+    return required_sections;
+  }
+}
+
+module.exports = QuestionnaireFlow;
```
Comment (MattElcock) [inline src/eq_schema/schema/QuestionnaireFlow/index.js:32]: You've still got some snake case in here that should be camel case.

For instance,

- Variable names, such as `const required_sections` and `const required_completed_sections`
- Functions, such as `build_options` and `build_required_sections`

File: src/eq_schema/schema/QuestionnaireFlow/index.test.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
+const questionnaire_flow = require(".");
```
Comment (MattElcock) [inline src/eq_schema/schema/QuestionnaireFlow/index.test.js:1]:
```suggestion
const QuestionnaireFlow = require(".");
```

This should also be camel case, and classes should start with a capital first letter.

**Review by MattElcock — CHANGES_REQUESTED** [review]

File: src/eq_schema/schema/QuestionnaireFlow/index.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    this.options = this.buildOptions(questionnaireJson);
+  }
+
+  buildOptions(questionnaireJson) {
+    const options = {}
+    if(questionnaireJson.hub) {
+      const requiredCompletedSections =
+        this.buildRequiredSections(questionnaireJson);
+      if (requiredCompletedSections.length) {
+        options.requiredCompletedSections = requiredCompletedSections;
```
Comment (MattElcock) [inline src/eq_schema/schema/QuestionnaireFlow/index.js:14]:
```suggestion
        options.required_completed_sections = requiredCompletedSections;
```

`options.required_completed_sections` is part of Runner's JSON, so should be snake case.

**Review by MattElcock — CHANGES_REQUESTED** [review]

**Review by MattElcock — APPROVED** [review]

**Review by martyncolmer — APPROVED** [review]

**Pull request merged**